### PR TITLE
[MGDSTRM-8218] Implement non-breaking App Services API guidelines for admin server

### LINKED ISSUE #127: Error responses should return a unique error code, different to the HTTP status code
The Kafka Admin API returns an HTTP status code, error message, and the error class as the main properties of the response payload when there is an error.

Clients consuming the API needs to be able to differentiate between different error types which are grouped under the same status code.

Example: when I try to reset the offset of a consumer group, I can get a `400 Bad Request` in various scenarios. The client will probably want to display a custom error message to the user, as sometimes the  response message is too technical for the user and for internationalisation purposes. Without the presence of a unique error code, I have no way to differentiate and appropriately handle this information in the client.

An example of an error response with this new payload format:

```json
{
   "code": 43982,
   "status":404,
   "error_message":"This server does not host this topic-partition.",
   "class":"UnknownTopicOrPartitionException"
}
```

To keep this backwards compatible, we could call the new error code `error_code` instead.

### PR DESCRIPTION
**Pull request by MikeEdgar**
Fixes #127

### CODE REVIEW

File: kafka-admin/.openapi/kafka-admin-rest.yaml [hunk 1]
```diff
@@ -792,6 +794,45 @@ paths:
           $ref: '#/components/responses/NotAuthorized'
         "500":
           $ref: '#/components/responses/ServerError'
+  /api/v1/errors:
```
Comment (wtrocki) [inline kafka-admin/.openapi/kafka-admin-rest.yaml:797]: Question (I did not run it yet) is that accessible without authentication?

Comment (MikeEdgar) [reply kafka-admin/.openapi/kafka-admin-rest.yaml:797]: Yes, there is no authentication. I'll make sure that's represented in the API definition.

**Comment by wtrocki** [comment]
@MikeEdgar would you have response from errors endpoint handy?

**Comment by MikeEdgar** [comment]
Here are a couple examples. One without `detail`, which is a property added for a specific occurrence of an error, but not returned from the general endpoint. Note that `code` is unchanged and still contains the HTTP status. Once we move to a `v2` set of endpoints that can be addressed (along with other breaking changes).
Request (numPartitions is a string)
```
curl localhost:8080/rest/topics --data '{ "name": "t2", "settings": { "numPartitions": "three" } }' -H"Content-type:application/json"
```
Response
```json
{
  "id": "7",
  "kind": "Error",
  "href": "/api/v1/errors/7",
  "reason": "Request body is malformed or invalid",
  "code": 400,
  "error_message": "Request body is malformed or invalid"
}
```
Request (numPartitions > limit)
```
curl localhost:8080/rest/topics --data '{ "name": "t2", "settings": { "numPartitions": "300" } }' -H"Content-type:application/json"
```
Response
```json
{
  "id": "16",
  "kind": "Error",
  "href": "/api/v1/errors/16",
  "reason": "Request body or parameters invalid",
  "detail": "numPartitions must be between 1 and 100 (inclusive)",
  "code": 400,
  "error_message": "Request body or parameters invalid"
}
```

**Comment by wtrocki** [comment]
Really sorry for not being clear. I meant result of 

```
curl <yourserver>/api/v1/errors
```

**Comment by rkpattnaik780** [comment]
Getting few errors after generating client in Go SDK, wherever `items` has type `[]map[string]interface{}`

```
cannot use items (variable of type []map[string]interface{}) as *[]AclBinding value in assignment
cannot use items (variable of type []map[string]interface{}) as *[]ConsumerGroupResetOffsetResultItem value in assignment
```

**Comment by wtrocki** [comment]
@rkpattnaik780 do we know what part of OpenAPI file causes that?

**Comment by rkpattnaik780** [comment]
> > Getting few errors after generating client in Go SDK, wherever `items` has type `[]map[string]interface{}`
> 
> Is that the equivalent type for the general list items? https://github.com/bf2fc6cc711aee1a0c2a/kafka-admin-api/pull/200/files#diff-28a7ff5a0ccb4c306af5199105457eb5c1547945322341cdcac828b585e09f5cR1427-R1431

That seems to be it.

**Comment by rkpattnaik780** [comment]
> Getting few errors after generating client in Go SDK, wherever `items` has type `[]map[string]interface{}`
> 
> ```
> cannot use items (variable of type []map[string]interface{}) as *[]AclBinding value in assignment
> cannot use items (variable of type []map[string]interface{}) as *[]ConsumerGroupResetOffsetResultItem value in assignment
> ```

@wtrocki I am seeing these errors in local, build is passing in PR check.

**Comment by wtrocki** [comment]
@rkpattnaik780 on go SDK we do not use examples so some errors will be seen only locally

**Comment by rkpattnaik780** [comment]
For the changes in [model_acl_binding_list_page.go](https://github.com/redhat-developer/app-services-sdk-go/pull/450/files#diff-97e01b0b6dbe47a5ab25fdb19952b18345e21f77715ac57678fc142de624f867L37-L40). We get the following error:
```
cannot use items (variable of type []map[string]interface{}) as *[]AclBinding value in assignment
```

Similarly for [model_record_list.go](https://github.com/redhat-developer/app-services-sdk-go/pull/450/files#diff-6fde3b442ce8aada75d04685464cbff184255fab32591ce63b205f3e1ab597c4L37-L42), the error is:
```
cannot use items (variable of type []map[string]interface{}) as *[]Record value in assignment
```

**Comment by wtrocki** [comment]
I think all SDKs somehow use there hashmap as input instead of object.
However the only golang is failing on this.

I'm not sure why `[]map[string]interface{}` is used for ACLbinding.
Did we changed anything in this PR that might have affected it?

Java PR
https://github.com/redhat-developer/app-services-sdk-java/pull/285

**Comment by MikeEdgar** [comment]
Yes - it most likely has something to do with the use of the new `List` schema. It's done similarly to other APIs that I've seen except that the `items` of `List` are general `object`s rather than `ObjectReference`s.

File: kafka-admin/.openapi/kafka-admin-rest.yaml [hunk 2]
```diff
@@ -1146,6 +1186,8 @@ components:
       - groupId
       - consumers
       type: object
+      allOf:
```
Comment (wtrocki) [inline kafka-admin/.openapi/kafka-admin-rest.yaml:1189]: This is the source of the problem. 
It doesn't seem to look correct to me:

allOff: is aggregation mechanism to merge multiple objects. 
When used it should aggregate all properties/objects. 
In current form it is only containing single object - meaning that it is redundant.

Additionally when using allOff we should have all properties defined inside it. 
Having properties outside allOff,oneOff etc. creates non deterministic schema 
that fails under each individual generator implementation.
What I see is that most of the generators give up in this case and simply does use HashMap.
However golang generator has bug for that use case and still expects proper object.

**Review by wtrocki — APPROVED** [review]

**Comment by wtrocki** [comment]
Good to merge.

**Comment by wtrocki** [comment]
Changes were integrated. I have tested this with Java (operator) and Golang and things are working fine. We can merge this

**Comment by MikeEdgar** [comment]
> Changes were integrated. I have tested this with Java (operator) and Golang and things are working fine. We can merge this

@wtrocki I'd like to push one more commit with the OpenAPI version set to 0.11.0 (the release this will be included in). Is that going to cause more work on your side?

**Comment by wtrocki** [comment]
No problem. That is actually very good to have. We will pick that up thru automatic changes to fully verify it.

**Comment by MikeEdgar** [comment]
@wtrocki this now includes the discussed changes to use `allOf` for all occurrences of inheritance. I'll merge so that the SDKs get the update. Please let me know if anything arises in the SDK generation that is not backward-compatible.

**Pull request merged**